File: sdk-python/obsvr/strict_receipt_coordinator_v2_1_recovery.py
"""Recoverable profile-2.1 coordinator with authenticated checkpoints."""

from __future__ import annotations

import copy
from typing import Any, Dict, Optional

from .strict_receipt_coordinator_v2_1 import StrictReceiptCoordinatorV21
from .strict_receipt_coordinator_v2_1_support import normalize_decision_action_v2_1
from .strict_receipt_reconcile_v2_1 import assert_accepted_strict_reconciliation_v2_1
from .strict_receipt_recovery_v2_1 import (
    sign_strict_recovery_v2_1,
    verify_strict_recovery_v2_1,
)


class RecoverableStrictReceiptCoordinatorV21(StrictReceiptCoordinatorV21):
    """Restore a prepared decision, reconcile it, and never invoke an action."""

    def __init__(
        self,
        *,
        sdk_version: str,
        recovery_checkpoint: Optional[Dict[str, Any]] = None,
        expected_origin_pid: Optional[int] = None,
        **options: Any,
    ) -> None:
        super().__init__(**options)
        if not isinstance(sdk_version, str) or not sdk_version.strip():
            raise ValueError("sdk_version must be nonblank")
        self._sdk_version = sdk_version
        self._recovery_prepared: Optional[Dict[str, Any]] = None
        if recovery_checkpoint is not None:
            document = verify_strict_recovery_v2_1(
                recovery_checkpoint, options["signer"]
            )
            self._restore(document, expected_origin_pid)

    def prepare_decision(self, input_value: Dict[str, Any]) -> Dict[str, Any]:
        with self._lock:
            self._assert_recovery_ready()
            normalized = normalize_decision_action_v2_1(input_value)
            prepared = super().prepare_decision(normalized)
            self._recovery_prepared = {
                "kind": "decision",
                "input": copy.deepcopy(normalized),
                "result": copy.deepcopy(prepared["value"]),
            }
            return prepared

    def prepare_approval_resolution(
        self, input_value: Dict[str, Any]
    ) -> Dict[str, Any]:
        with self._lock:
            self._assert_recovery_ready()
            prepared = super().prepare_approval_resolution(input_value)
            self._recovery_prepared = {
                "kind": "resolution",
                "suspended_receipt_hash": prepared["value"]["body"]["resolution"][
                    "resolves_receipt_hash"
                ],
                "result": copy.deepcopy(prepared["value"]),
            }
            return prepared

    def commit_prepared(self, token: str, receipt_hash: str) -> Dict[str, Any]:
        result = super().commit_prepared(token, receipt_hash)
        self._recovery_prepared = None
        return result

    def abort_prepared(self, token: str, receipt_hash: str, capability: Any) -> None:
        super().abort_prepared(token, receipt_hash, capability)
        self._recovery_prepared = None

    def inspect_state(self) -> Dict[str, Any]:
        state = super().inspect_state()
        if self._recovery_prepared is not None and "prepared" not in state:
            state.update(frozen=True, freeze_reason="restart_reconciliation_required")
        return state

    def export_recovery_checkpoint(self) -> Dict[str, Any]:
        with self._lock:
            self._ensure_process()
            document = {
                "schema": "obsvr-strict-receipt-recovery-v2-1",
                "profile_version": "2.1",
                "tenant_id": self._tenant_id,
                "session_id": self._session_id,
                "sdk_language": "python",
                "sdk_version": self._sdk_version,
                "origin_pid": self._owner_pid,
                "committed": {
                    "sequence": self._sequence,
                    "head_receipt_hash": self._head_receipt_hash,
                    "last_timestamp_ms": self._last_timestamp,
                    "prior_actions": copy.deepcopy(self._prior_actions),
                    "action_ids": sorted(self._committed_action_ids),
                    "pending_approval_ids": sorted(self._pending_approval_ids),
                    "suspended_approvals": [
                        copy.deepcopy(self._suspended_approvals[key])
                        for key in sorted(self._suspended_approvals)
                    ],
                    "resolved_approval_hashes": sorted(self._resolved_approvals),
                },
            }
            if self._recovery_prepared is not None:
                document["prepared"] = copy.deepcopy(self._recovery_prepared)
            return sign_strict_recovery_v2_1(document, self._signer)

    def reconcile_recovered_accepted(self, proof: Any) -> Dict[str, Any]:
        with self._lock:
            self._ensure_process()
            pending = self._recovery_prepared
            if pending is None:
                raise ValueError("no recovered decision is pending")
            receipt = (
                pending["result"]["receipt"]
                if pending["kind"] == "decision"
                else pending["result"]
            )
            assert_accepted_strict_reconciliation_v2_1(proof, receipt)
            if pending["kind"] == "decision":
                self._commit_decision(pending["result"], pending["input"])
            else:
                target_hash = pending["suspended_receipt_hash"]
                index = next(
                    (
                        index
                        for index, item in enumerate(self._prior_actions)
                        if item["receipt_hash"] == target_hash
                    ),
                    -1,
                )
                if index < 0:
                    raise ValueError("recovered approval action summary is missing")
                self._commit_approval_resolution(receipt, target_hash, index)
            self._recovery_prepared = None
            return copy.deepcopy(receipt)

    def _restore(self, document: Dict[str, Any], expected_origin_pid: Any) -> None:
        if (
            isinstance(expected_origin_pid, bool)
            or not isinstance(expected_origin_pid, int)
            or expected_origin_pid < 0
            or document["origin_pid"] != expected_origin_pid
        ):
            raise ValueError("recovery origin PID mismatch")
        if (
            document["tenant_id"] != self._tenant_id
            or document["session_id"] != self._session_id
            or document["sdk_language"] != "python"
            or document["sdk_version"] != self._sdk_version
            or document["profile_version"] != "2.1"
        ):
            raise ValueError("recovery tenant/session/sdk/profile mismatch")
        state = document["committed"]
        self._sequence = state["sequence"]
        self._head_receipt_hash = state["head_receipt_hash"]
        self._last_timestamp = state["last_timestamp_ms"]
        self._prior_actions = copy.deepcopy(state["prior_actions"])
        self._committed_action_ids = set(state["action_ids"])
        self._pending_approval_ids = set(state["pending_approval_ids"])
        self._suspended_approvals = {
            item["receipt"]["receipt_hash"]: copy.deepcopy(item)
            for item in state["suspended_approvals"]
        }
        self._resolved_approvals = set(state["resolved_approval_hashes"])
        self._recovery_prepared = copy.deepcopy(document.get("prepared"))

    def _assert_recovery_ready(self) -> None:
        if self._recovery_prepared is not None:
            raise ValueError(
                "recovered decision requires accepted reconciliation before new work"
            )


__all__ = ["RecoverableStrictReceiptCoordinatorV21"]

File: sdk-typescript/src/governance/strict-receipt-coordinator-v2-1-recovery.ts
import {
  StrictReceiptCoordinatorV21,
} from './strict-receipt-coordinator-v2-1.js';
import {
  normalizeDecisionActionV21, v21Clone,
} from './strict-receipt-coordinator-v2-1-support.js';
import type {
  PreparedDecisionV21, StrictCoordinatorV21StateInspection,
  StrictDecisionActionV21Input, StrictDecisionV21Result,
  StrictReceiptCoordinatorV21Options,
} from './strict-receipt-coordinator-v2-1-types.js';
import type { DefinitiveNoStore } from './strict-receipt-prepared-state.js';
import type { StrictReceiptV21Envelope } from './strict-receipt-v2-1.js';
import {
  signStrictRecoveryV21, type RecoveryPreparedV21,
  type StrictRecoveryV21Document, type StrictRecoveryV21Envelope,
  verifyStrictRecoveryV21,
} from './strict-receipt-recovery-v2-1.js';
import {
  assertAcceptedStrictReconciliationV21, type StrictReconciliationV21Result,
} from './strict-receipt-reconcile-v2-1.js';

export interface RecoverableStrictReceiptCoordinatorV21Options
  extends StrictReceiptCoordinatorV21Options { sdk_version: string }
export interface StrictCoordinatorV21Restore {
  checkpoint: StrictRecoveryV21Envelope;
  expected_origin_pid: number;
}

export class RecoverableStrictReceiptCoordinatorV21 extends StrictReceiptCoordinatorV21 {
  private readonly sdkVersion: string;
  private recoveryPrepared: RecoveryPreparedV21 | undefined;

  constructor(options: RecoverableStrictReceiptCoordinatorV21Options, restore?: StrictCoordinatorV21Restore) {
    super(options);
    if (typeof options.sdk_version !== 'string' || !options.sdk_version.trim()) throw new Error('sdk_version must be nonblank');
    this.sdkVersion = options.sdk_version;
    if (restore) this.restore(
      verifyStrictRecoveryV21(restore.checkpoint, options.signer), restore.expected_origin_pid,
    );
  }

  override prepareDecision(input: StrictDecisionActionV21Input): PreparedDecisionV21 {
    this.assertRecoveryReady();
    const normalized = normalizeDecisionActionV21(input);
    const prepared = super.prepareDecision(normalized);
    this.recoveryPrepared = { kind: 'decision', input: v21Clone(normalized), result: v21Clone(prepared.value) };
    return prepared;
  }

  override commitPrepared(token: string, hash: string): StrictDecisionV21Result {
    const result = super.commitPrepared(token, hash); this.recoveryPrepared = undefined; return result;
  }

  override abortPrepared(token: string, hash: string, capability: DefinitiveNoStore): void {
    super.abortPrepared(token, hash, capability); this.recoveryPrepared = undefined;
  }

  override inspectState(): StrictCoordinatorV21StateInspection {
    const state = super.inspectState();
    return this.recoveryPrepared && !state.prepared
      ? { ...state, frozen: true, freeze_reason: 'restart_reconciliation_required' }
      : state;
  }

  exportRecoveryCheckpoint(): StrictRecoveryV21Envelope {
    this.ensureProcess();
    const document: StrictRecoveryV21Document = {
      schema: 'obsvr-strict-receipt-recovery-v2-1', profile_version: '2.1',
      tenant_id: this.tenantId, session_id: this.sessionId,
      sdk_language: 'typescript', sdk_version: this.sdkVersion, origin_pid: this.ownerPid,
      committed: {
        sequence: this.sequence, head_receipt_hash: this.headReceiptHash,
        last_timestamp_ms: this.lastTimestamp, prior_actions: v21Clone(this.priorActions),
        action_ids: [...this.committedActionIds].sort(),
        pending_approval_ids: [...this.pendingApprovalIds].sort(),
      },
      ...(this.recoveryPrepared ? { prepared: v21Clone(this.recoveryPrepared) } : {}),
    };
    return signStrictRecoveryV21(document, this.options.signer);
  }

  reconcileRecoveredAccepted(proof: StrictReconciliationV21Result): StrictReceiptV21Envelope {
    this.ensureProcess(); const pending = this.recoveryPrepared;
    if (!pending) throw new Error('no recovered decision is pending');
    assertAcceptedStrictReconciliationV21(proof, pending.result.receipt);
    this.commitDecision(pending.result, pending.input);
    this.recoveryPrepared = undefined;
    return v21Clone(pending.result.receipt);
  }

  private restore(document: StrictRecoveryV21Document, expectedOriginPid: number): void {
    if (!Number.isSafeInteger(expectedOriginPid) || expectedOriginPid < 0
      || document.origin_pid !== expectedOriginPid) throw new Error('recovery origin PID mismatch');
    if (document.tenant_id !== this.tenantId || document.session_id !== this.sessionId
      || document.sdk_language !== 'typescript' || document.sdk_version !== this.sdkVersion
      || document.profile_version !== '2.1') throw new Error('recovery tenant/session/sdk/profile mismatch');
    const state = document.committed;
    this.sequence = state.sequence; this.headReceiptHash = state.head_receipt_hash;
    this.lastTimestamp = state.last_timestamp_ms; this.priorActions = v21Clone(state.prior_actions);
    this.committedActionIds.clear(); state.action_ids.forEach((id) => this.committedActionIds.add(id));
    this.pendingApprovalIds.clear(); state.pending_approval_ids.forEach((id) => this.pendingApprovalIds.add(id));
    this.recoveryPrepared = document.prepared ? v21Clone(document.prepared) : undefined;
  }

  private assertRecoveryReady(): void {
    if (this.recoveryPrepared) throw new Error('recovered decision requires accepted reconciliation before new work');
  }
}
